File: codealg.py
import random
import time
import statistics
import math
import csv
import matplotlib.pyplot as plt

# ==========================================
# FUNCIONES GENERADORAS DE EXPERIMENTOS
# ==========================================
def generar_exp1_enteros(n):
    A = [[random.randint(1, 10) for _ in range(n)] for _ in range(n)]
    B = [[random.randint(1, 10) for _ in range(n)] for _ in range(n)]
    return A, B

def generar_exp2_reales_0_1(n):
    A = [[random.random() for _ in range(n)] for _ in range(n)]
    B = [[random.random() for _ in range(n)] for _ in range(n)]
    return A, B

def generar_exp3_estructuradas(n):
    A = [[(i + j + 1) for j in range(n)] for i in range(n)]
    B = [[(i - j) for j in range(n)] for i in range(n)]
    return A, B

def generar_exp4_reales_2_decimales(n):
    A = [[round(random.uniform(1.0, 100.0), 2) for _ in range(n)] for _ in range(n)]
    B = [[round(random.uniform(1.0, 100.0), 2) for _ in range(n)] for _ in range(n)]
    return A, B

def generar_exp5_longlong(n):
    min_val, max_val = 10**15, 10**16
    A = [[random.randint(min_val, max_val) for _ in range(n)] for _ in range(n)]
    B = [[random.randint(min_val, max_val) for _ in range(n)] for _ in range(n)]
    return A, B

# ==========================================
# ALGORITMOS MATEMÁTICOS
# ==========================================
def multiplicacion_clasica(A, B):
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C

def sumar_matrices(A, B):
    n = len(A)
    return [[A[i][j] + B[i][j] for j in range(n)] for i in range(n)]

def restar_matrices(A, B):
    n = len(A)
    return [[A[i][j] - B[i][j] for j in range(n)] for i in range(n)]

def dividir_matriz(A):
    n = len(A)
    mid = n // 2
    return [fila[:mid] for fila in A[:mid]], [fila[mid:] for fila in A[:mid]], \
           [fila[:mid] for fila in A[mid:]], [fila[mid:] for fila in A[mid:]]

def multiplicacion_strassen(A, B):
    n = len(A)
    if n == 1:
        return [[A[0][0] * B[0][0]]]
    
    A11, A12, A21, A22 = dividir_matriz(A)
    B11, B12, B21, B22 = dividir_matriz(B)
    
    S1 = sumar_matrices(A11, A22); S2 = sumar_matrices(B11, B22)
    S3 = sumar_matrices(A21, A22); S4 = restar_matrices(B12, B22)
    S5 = restar_matrices(B21, B11); S6 = sumar_matrices(A11, A12)
    S7 = restar_matrices(A21, A11); S8 = sumar_matrices(B11, B12)
    S9 = restar_matrices(A12, A22); S10 = sumar_matrices(B21, B22)
    
    M1 = multiplicacion_strassen(S1, S2); M2 = multiplicacion_strassen(S3, B11)
    M3 = multiplicacion_strassen(A11, S4); M4 = multiplicacion_strassen(A22, S5)
    M5 = multiplicacion_strassen(S6, B22); M6 = multiplicacion_strassen(S7, S8)
    M7 = multiplicacion_strassen(S9, S10)
    
    C11 = sumar_matrices(restar_matrices(sumar_matrices(M1, M4), M5), M7)
    C12 = sumar_matrices(M3, M5)
    C21 = sumar_matrices(M2, M4)
    C22 = sumar_matrices(sumar_matrices(restar_matrices(M1, M2), M3), M6)
    
    C = [[0 for _ in range(n)] for _ in range(n)]
    mid = n // 2
    for i in range(mid):
        for j in range(mid):
            C[i][j] = C11[i][j]; C[i][j + mid] = C12[i][j]
            C[i + mid][j] = C21[i][j]; C[i + mid][j + mid] = C22[i][j]
    return C

def multiplicacion_hibrida(A, B, n0):
    n = len(A)
    if n <= n0:
        return multiplicacion_clasica(A, B)
    
    A11, A12, A21, A22 = dividir_matriz(A)
    B11, B12, B21, B22 = dividir_matriz(B)
    
    S1 = sumar_matrices(A11, A22); S2 = sumar_matrices(B11, B22)
    S3 = sumar_matrices(A21, A22); S4 = restar_matrices(B12, B22)
    S5 = restar_matrices(B21, B11); S6 = sumar_matrices(A11, A12)
    S7 = restar_matrices(A21, A11); S8 = sumar_matrices(B11, B12)
    S9 = restar_matrices(A12, A22); S10 = sumar_matrices(B21, B22)
    
    M1 = multiplicacion_hibrida(S1, S2, n0); M2 = multiplicacion_hibrida(S3, B11, n0)
    M3 = multiplicacion_hibrida(A11, S4, n0); M4 = multiplicacion_hibrida(A22, S5, n0)
    M5 = multiplicacion_hibrida(S6, B22, n0); M6 = multiplicacion_hibrida(S7, S8, n0)
    M7 = multiplicacion_hibrida(S9, S10, n0)
    
    C11 = sumar_matrices(restar_matrices(sumar_matrices(M1, M4), M5), M7)
    C12 = sumar_matrices(M3, M5)
    C21 = sumar_matrices(M2, M4)
    C22 = sumar_matrices(sumar_matrices(restar_matrices(M1, M2), M3), M6)
    
    C = [[0 for _ in range(n)] for _ in range(n)]
    mid = n // 2
    for i in range(mid):
        for j in range(mid):
            C[i][j] = C11[i][j]; C[i][j + mid] = C12[i][j]
            C[i + mid][j] = C21[i][j]; C[i + mid][j + mid] = C22[i][j]
    return C

# ==========================================
# RUTINA PRINCIPAL DE BENCHMARKING
# ==========================================
if __name__ == "__main__":
    tamanos_n = [16, 32, 64, 128, 256, 512]
    repeticiones = 32
    umbrales_n0_a_probar = [16, 32, 64] # Umbrales a evaluar analíticamente
    
    experimentos = {
        "Exp1_EnterosAleatorios": generar_exp1_enteros,
        "Exp2_Reales_0a1": generar_exp2_reales_0_1,
        "Exp3_Estructuradas": generar_exp3_estructuradas,
        "Exp4_Reales2Decimales": generar_exp4_reales_2_decimales,
        "Exp5_LongLong": generar_exp5_longlong
    }

    print("--- INICIANDO BENCHMARKING ---")
    resultados_globales = {}
    
    for nombre_exp, funcion_generadora in experimentos.items():
        print(f"\nEjecutando {nombre_exp}")
        is_hibrido_target = (nombre_exp == "Exp5_LongLong")
        
        resultados_globales[nombre_exp] = {
            'n': [], 't_clasico': [], 'std_clasico': [], 't_strassen': [], 'std_strassen': []
        }
        
        if is_hibrido_target:
            for n0 in umbrales_n0_a_probar:
                resultados_globales[nombre_exp][f't_hib_{n0}'] = []
                resultados_globales[nombre_exp][f'std_hib_{n0}'] = []
        
        for n in tamanos_n:
            print(f"  Midiendo n={n} ({repeticiones} reps)...")
            tiempos_clasico, tiempos_strassen = [], []
            tiempos_hibrido = {n0: [] for n0 in umbrales_n0_a_probar}
            
            for _ in range(repeticiones):
                A, B = funcion_generadora(n)
                
                # Clásico
                t_ini = time.perf_counter()
                multiplicacion_clasica(A, B)
                tiempos_clasico.append(time.perf_counter() - t_ini)
                
                # Strassen
                t_ini = time.perf_counter()
                multiplicacion_strassen(A, B)
                tiempos_strassen.append(time.perf_counter() - t_ini)
                
                # Híbrido (SÓLO SI ES EL EXPERIMENTO LONG LONG)
                if is_hibrido_target:
                    for n0 in umbrales_n0_a_probar:
                        t_ini = time.perf_counter()
                        multiplicacion_hibrida(A, B, n0)
                        tiempos_hibrido[n0].append(time.perf_counter() - t_ini)
                
            resultados_globales[nombre_exp]['n'].append(n)
            resultados_globales[nombre_exp]['t_clasico'].append(statistics.mean(tiempos_clasico))
            resultados_globales[nombre_exp]['std_clasico'].append(statistics.stdev(tiempos_clasico) if repeticiones > 1 else 0)
            resultados_globales[nombre_exp]['t_strassen'].append(statistics.mean(tiempos_strassen))
            resultados_globales[nombre_exp]['std_strassen'].append(statistics.stdev(tiempos_strassen) if repeticiones > 1 else 0)
            
            if is_hibrido_target:
                for n0 in umbrales_n0_a_probar:
                    resultados_globales[nombre_exp][f't_hib_{n0}'].append(statistics.mean(tiempos_hibrido[n0]))
                    resultados_globales[nombre_exp][f'std_hib_{n0}'].append(statistics.stdev(tiempos_hibrido[n0]) if repeticiones > 1 else 0)

    # CREACIÓN DEL ARCHIVO CSV
    print("\nGenerando CSV (resultados_python.csv)...")
    with open('resultados_python.csv', mode='w', newline='') as file:
        writer = csv.writer(file)
        headers = ['Experimento', 'Tamano_n', 'Prom_Clasico', 'Std_Clasico', 'Prom_Strassen', 'Std_Strassen']
        headers += [f'Prom_Hibrido(n0={n0})' for n0 in umbrales_n0_a_probar]
        headers += [f'Std_Hibrido(n0={n0})' for n0 in umbrales_n0_a_probar]
        writer.writerow(headers)
        
        for nombre_exp, datos in resultados_globales.items():
            is_hibrido_target = (nombre_exp == "Exp5_LongLong")
            for i in range(len(datos['n'])):
                row = [
                    nombre_exp, datos['n'][i],
                    f"{datos['t_clasico'][i]:.6f}", f"{datos['std_clasico'][i]:.6f}",
                    f"{datos['t_strassen'][i]:.6f}", f"{datos['std_strassen'][i]:.6f}"
                ]
                
                if is_hibrido_target:
                    for n0 in umbrales_n0_a_probar:
                        row.append(f"{datos[f't_hib_{n0}'][i]:.6f}")
                    for n0 in umbrales_n0_a_probar:
                        row.append(f"{datos[f'std_hib_{n0}'][i]:.6f}")
                else:
                    row += [''] * (len(umbrales_n0_a_probar) * 2) # Celdas vacías para los otros exp
                    
                writer.writerow(row)

    # GRAFICACIÓN LOG-LOG
    print("Generando gráficos...")
    for nombre_exp, datos in resultados_globales.items():
        is_hibrido_target = (nombre_exp == "Exp5_LongLong")
        ns = datos['n']
        plt.figure(figsize=(10, 6))
        
        plt.loglog(ns, datos['t_clasico'], 'o-', linewidth=2, label='Clásico Empírico')
        plt.loglog(ns, datos['t_strassen'], 's-', linewidth=2, label='Strassen Puro')
        
        if is_hibrido_target:
            markers = ['^-', 'v-', 'd-']
            for idx, n0 in enumerate(umbrales_n0_a_probar):
                plt.loglog(ns, datos[f't_hib_{n0}'], markers[idx], linewidth=2, label=f'Híbrido (n0={n0})')
        
        c_clas = datos['t_clasico'][-1] / (ns[-1] ** 3)
        plt.loglog(ns, [c_clas * (n ** 3) for n in ns], 'k--', alpha=0.5, label='Teórico O(n^3)')
        
        c_stras = datos['t_strassen'][-1] / (ns[-1] ** 2.807)
        plt.loglog(ns, [c_stras * (n ** 2.807) for n in ns], 'r--', alpha=0.5, label='Teórico O(n^2.81)')
        
        plt.xlabel('Tamaño de Matriz (n) [Escala Log]')
        plt.ylabel('Tiempo Promedio (segundos) [Escala Log]')
        plt.title(f'Comparación de Algoritmos - {nombre_exp}')
        plt.xticks(ns, [str(val) for val in ns])
        plt.legend()
        plt.grid(True, which="both", ls="--", alpha=0.5)
        
        plt.savefig(f'Python_Grafico_{nombre_exp}.png')
        plt.close()
        
    print("Finalizado con éxito.")

File: codealg.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.IO;
using System.Text;
using System.Globalization;

namespace MultiplicacionMatrices
{
    class ResultadoExperimento
    {
        public string Nombre;
        public bool EsExpHibrido;
        public List<int> Ns = new List<int>();
        public List<double> TClasico = new List<double>();
        public List<double> StdClasico = new List<double>();
        public List<double> TStrassen = new List<double>();
        public List<double> StdStrassen = new List<double>();
        
        // Listas para los 3 umbrales a analizar analíticamente
        public List<double> THib16 = new List<double>(); public List<double> StdHib16 = new List<double>();
        public List<double> THib32 = new List<double>(); public List<double> StdHib32 = new List<double>();
        public List<double> THib64 = new List<double>(); public List<double> StdHib64 = new List<double>();
    }

    class Program
    {
        static Random rand = new Random();

        // Funciones generadoras
        static void GenExp1_Enteros(int n, out double[][] A, out double[][] B)
        {
            A = CreateMatrix(n); B = CreateMatrix(n);
            for (int i=0; i<n; i++) for (int j=0; j<n; j++) {
                A[i][j] = rand.Next(1, 11); B[i][j] = rand.Next(1, 11);
            }
        }

        static void GenExp2_Reales_0_1(int n, out double[][] A, out double[][] B)
        {
            A = CreateMatrix(n); B = CreateMatrix(n);
            for (int i=0; i<n; i++) for (int j=0; j<n; j++) {
                A[i][j] = rand.NextDouble(); B[i][j] = rand.NextDouble();
            }
        }

        static void GenExp3_Estructuradas(int n, out double[][] A, out double[][] B)
        {
            A = CreateMatrix(n); B = CreateMatrix(n);
            for (int i=0; i<n; i++) for (int j=0; j<n; j++) {
                A[i][j] = (double)(i + j + 1); B[i][j] = (double)(i - j);
            }
        }

        static void GenExp4_Reales2Decimales(int n, out double[][] A, out double[][] B)
        {
            A = CreateMatrix(n); B = CreateMatrix(n);
            for (int i=0; i<n; i++) for (int j=0; j<n; j++) {
                A[i][j] = Math.Round(rand.NextDouble() * 100.0, 2); B[i][j] = Math.Round(rand.NextDouble() * 100.0, 2);
            }
        }

        static void GenExp5_LongLong(int n, out double[][] A, out double[][] B)
        {
            A = CreateMatrix(n); B = CreateMatrix(n);
            double min = 100000000000000.0; double max = 900000000000000.0;
            for (int i=0; i<n; i++) for (int j=0; j<n; j++) {
                A[i][j] = Math.Floor(min + (rand.NextDouble() * (max - min))); B[i][j] = Math.Floor(min + (rand.NextDouble() * (max - min)));
            }
        }

        static double[][] CreateMatrix(int n)
        {
            double[][] result = new double[n][];
            for (int i = 0; i < n; i++) result[i] = new double[n];
            return result;
        }

        static double CalcularDesviacionEstandar(List<double> valores, double promedio)
        {
            if (valores.Count <= 1) return 0.0;
            double sumOfSquares = valores.Sum(val => Math.Pow(val - promedio, 2));
            return Math.Sqrt(sumOfSquares / (valores.Count - 1));
        }

        // ==========================================
        // ALGORITMOS MATEMÁTICOS
        // ==========================================
        static double[][] MultiplicacionClasica(double[][] A, double[][] B)
        {
            int n = A.Length;
            double[][] C = CreateMatrix(n);
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++)
                    for (int k = 0; k < n; k++)
                        C[i][j] += A[i][k] * B[k][j];
            return C;
        }

        static double[][] SumarMatrices(double[][] A, double[][] B)
        {
            int n = A.Length;
            double[][] C = CreateMatrix(n);
            for (int i = 0; i < n; i++) for (int j = 0; j < n; j++) C[i][j] = A[i][j] + B[i][j];
            return C;
        }

        static double[][] RestarMatrices(double[][] A, double[][] B)
        {
            int n = A.Length;
            double[][] C = CreateMatrix(n);
            for (int i = 0; i < n; i++) for (int j = 0; j < n; j++) C[i][j] = A[i][j] - B[i][j];
            return C;
        }

        static void DividirMatriz(double[][] A, out double[][] A11, out double[][] A12, out double[][] A21, out double[][] A22)
        {
            int mid = A.Length / 2;
            A11 = CreateMatrix(mid); A12 = CreateMatrix(mid);
            A21 = CreateMatrix(mid); A22 = CreateMatrix(mid);
            for (int i = 0; i < mid; i++) for (int j = 0; j < mid; j++) {
                A11[i][j] = A[i][j]; A12[i][j] = A[i][j + mid];
                A21[i][j] = A[i + mid][j]; A22[i][j] = A[i + mid][j + mid];
            }
        }

        static double[][] MultiplicacionStrassen(double[][] A, double[][] B)
        {
            int n = A.Length;
            if (n == 1) { double[][] C = CreateMatrix(1); C[0][0] = A[0][0] * B[0][0]; return C; }

            double[][] A11, A12, A21, A22, B11, B12, B21, B22;
            DividirMatriz(A, out A11, out A12, out A21, out A22); DividirMatriz(B, out B11, out B12, out B21, out B22);

            double[][] S1 = SumarMatrices(A11, A22); double[][] S2 = SumarMatrices(B11, B22);
            double[][] S3 = SumarMatrices(A21, A22); double[][] S4 = RestarMatrices(B12, B22);
            double[][] S5 = RestarMatrices(B21, B11); double[][] S6 = SumarMatrices(A11, A12);
            double[][] S7 = RestarMatrices(A21, A11); double[][] S8 = SumarMatrices(B11, B12);
            double[][] S9 = RestarMatrices(A12, A22); double[][] S10 = SumarMatrices(B21, B22);

            double[][] M1 = MultiplicacionStrassen(S1, S2); double[][] M2 = MultiplicacionStrassen(S3, B11);
            double[][] M3 = MultiplicacionStrassen(A11, S4); double[][] M4 = MultiplicacionStrassen(A22, S5);
            double[][] M5 = MultiplicacionStrassen(S6, B22); double[][] M6 = MultiplicacionStrassen(S7, S8);
            double[][] M7 = MultiplicacionStrassen(S9, S10);

            double[][] C11 = SumarMatrices(RestarMatrices(SumarMatrices(M1, M4), M5), M7);
            double[][] C12 = SumarMatrices(M3, M5);
            double[][] C21 = SumarMatrices(M2, M4);
            double[][] C22 = SumarMatrices(SumarMatrices(RestarMatrices(M1, M2), M3), M6);

            double[][] finalC = CreateMatrix(n); int mid = n / 2;
            for (int i = 0; i < mid; i++) for (int j = 0; j < mid; j++) {
                finalC[i][j] = C11[i][j]; finalC[i][j + mid] = C12[i][j];
                finalC[i + mid][j] = C21[i][j]; finalC[i + mid][j + mid] = C22[i][j];
            }
            return finalC;
        }

        static double[][] MultiplicacionHibrida(double[][] A, double[][] B, int n0)
        {
            int n = A.Length;
            if (n <= n0) return MultiplicacionClasica(A, B);

            double[][] A11, A12, A21, A22, B11, B12, B21, B22;
            DividirMatriz(A, out A11, out A12, out A21, out A22); DividirMatriz(B, out B11, out B12, out B21, out B22);

            double[][] S1 = SumarMatrices(A11, A22); double[][] S2 = SumarMatrices(B11, B22);
            double[][] S3 = SumarMatrices(A21, A22); double[][] S4 = RestarMatrices(B12, B22);
            double[][] S5 = RestarMatrices(B21, B11); double[][] S6 = SumarMatrices(A11, A12);
            double[][] S7 = RestarMatrices(A21, A11); double[][] S8 = SumarMatrices(B11, B12);
            double[][] S9 = RestarMatrices(A12, A22); double[][] S10 = SumarMatrices(B21, B22);

            double[][] M1 = MultiplicacionHibrida(S1, S2, n0); double[][] M2 = MultiplicacionHibrida(S3, B11, n0);
            double[][] M3 = MultiplicacionHibrida(A11, S4, n0); double[][] M4 = MultiplicacionHibrida(A22, S5, n0);
            double[][] M5 = MultiplicacionHibrida(S6, B22, n0); double[][] M6 = MultiplicacionHibrida(S7, S8, n0);
            double[][] M7 = MultiplicacionHibrida(S9, S10, n0);

            double[][] C11 = SumarMatrices(RestarMatrices(SumarMatrices(M1, M4), M5), M7);
            double[][] C12 = SumarMatrices(M3, M5);
            double[][] C21 = SumarMatrices(M2, M4);
            double[][] C22 = SumarMatrices(SumarMatrices(RestarMatrices(M1, M2), M3), M6);

            double[][] finalC = CreateMatrix(n); int mid = n / 2;
            for (int i = 0; i < mid; i++) for (int j = 0; j < mid; j++) {
                finalC[i][j] = C11[i][j]; finalC[i][j + mid] = C12[i][j];
                finalC[i + mid][j] = C21[i][j]; finalC[i + mid][j + mid] = C22[i][j];
            }
            return finalC;
        }

        delegate void GeneradorExperimento(int n, out double[][] A, out double[][] B);

        static void Main(string[] args)
        {
            int[] tamanosN = new int[] { 16, 32, 64, 128, 256, 512 };
            int repeticiones = 32;

            var experimentos = new Dictionary<string, GeneradorExperimento>();
            experimentos.Add("Exp1_Enteros", GenExp1_Enteros);
            experimentos.Add("Exp2_Reales_0a1", GenExp2_Reales_0_1);
            experimentos.Add("Exp3_Estructuradas", GenExp3_Estructuradas);
            experimentos.Add("Exp4_Reales2Dec", GenExp4_Reales2Decimales);
            experimentos.Add("Exp5_LongLong", GenExp5_LongLong);

            Stopwatch sw = new Stopwatch();
            List<ResultadoExperimento> listaResultados = new List<ResultadoExperimento>();

            Console.WriteLine("--- INICIANDO BENCHMARKING ---");

            foreach (var exp in experimentos)
            {
                Console.WriteLine(string.Format("=== Ejecutando {0} ===", exp.Key));
                ResultadoExperimento res = new ResultadoExperimento();
                res.Nombre = exp.Key;
                res.EsExpHibrido = (exp.Key == "Exp5_LongLong"); // SOLO ACTIVO EN LONG LONG

                foreach (int n in tamanosN)
                {
                    Console.Write(string.Format(" Midiendo tamano n={0}... ", n));
                    List<double> tiemposClasico = new List<double>();
                    List<double> tiemposStrassen = new List<double>();
                    List<double> tiemposHib16 = new List<double>(), tiemposHib32 = new List<double>(), tiemposHib64 = new List<double>();

                    for (int i = 0; i < repeticiones; i++)
                    {
                        double[][] A, B;
                        exp.Value(n, out A, out B);

                        sw.Restart(); MultiplicacionClasica(A, B); sw.Stop();
                        tiemposClasico.Add(sw.Elapsed.TotalSeconds);

                        sw.Restart(); MultiplicacionStrassen(A, B); sw.Stop();
                        tiemposStrassen.Add(sw.Elapsed.TotalSeconds);

                        // SI ES EL EXPERIMENTO LONG LONG, EVALÚA LOS UMBRALES HÍBRIDOS
                        if (res.EsExpHibrido)
                        {
                            sw.Restart(); MultiplicacionHibrida(A, B, 16); sw.Stop(); tiemposHib16.Add(sw.Elapsed.TotalSeconds);
                            sw.Restart(); MultiplicacionHibrida(A, B, 32); sw.Stop(); tiemposHib32.Add(sw.Elapsed.TotalSeconds);
                            sw.Restart(); MultiplicacionHibrida(A, B, 64); sw.Stop(); tiemposHib64.Add(sw.Elapsed.TotalSeconds);
                        }
                    }

                    res.Ns.Add(n);
                    
                    double pClas = tiemposClasico.Average();
                    res.TClasico.Add(pClas); res.StdClasico.Add(CalcularDesviacionEstandar(tiemposClasico, pClas));
                    
                    double pStras = tiemposStrassen.Average();
                    res.TStrassen.Add(pStras); res.StdStrassen.Add(CalcularDesviacionEstandar(tiemposStrassen, pStras));
                    
                    if (res.EsExpHibrido)
                    {
                        double pHib16 = tiemposHib16.Average(); res.THib16.Add(pHib16); res.StdHib16.Add(CalcularDesviacionEstandar(tiemposHib16, pHib16));
                        double pHib32 = tiemposHib32.Average(); res.THib32.Add(pHib32); res.StdHib32.Add(CalcularDesviacionEstandar(tiemposHib32, pHib32));
                        double pHib64 = tiemposHib64.Average(); res.THib64.Add(pHib64); res.StdHib64.Add(CalcularDesviacionEstandar(tiemposHib64, pHib64));
                    }
                    Console.WriteLine("Completado.");
                }
                listaResultados.Add(res);
            }

            Console.WriteLine("\nGenerando CSV (resultados_csharp.csv)...");
            CultureInfo ci = CultureInfo.InvariantCulture;
            using (StreamWriter file = new StreamWriter("resultados_csharp.csv"))
            {
                file.WriteLine("Experimento,Tamano_n,Prom_Clasico,Std_Clasico,Prom_Strassen,Std_Strassen,Prom_Hib(n0=16),Prom_Hib(n0=32),Prom_Hib(n0=64)");
                foreach (var res in listaResultados)
                {
                    for (int i = 0; i < res.Ns.Count; i++)
                    {
                        if (res.EsExpHibrido)
                        {
                            file.WriteLine(string.Format(ci, "{0},{1},{2:F6},{3:F6},{4:F6},{5:F6},{6:F6},{7:F6},{8:F6}", 
                                res.Nombre, res.Ns[i], res.TClasico[i], res.StdClasico[i], res.TStrassen[i], res.StdStrassen[i], res.THib16[i], res.THib32[i], res.THib64[i]));
                        }
                        else
                        {
                            file.WriteLine(string.Format(ci, "{0},{1},{2:F6},{3:F6},{4:F6},{5:F6},,,", 
                                res.Nombre, res.Ns[i], res.TClasico[i], res.StdClasico[i], res.TStrassen[i], res.StdStrassen[i]));
                        }
                    }
                }
            }

            Console.WriteLine("Generando graficos en Python...");
            StringBuilder pyCode = new StringBuilder();
            pyCode.AppendLine("import matplotlib.pyplot as plt\nimport sys\nexperimentos = {}");

            foreach (var res in listaResultados)
            {
                pyCode.AppendLine(string.Format("experimentos['{0}'] = {{}}", res.Nombre));
                pyCode.AppendLine(string.Format("experimentos['{0}']['n'] = [{1}]", res.Nombre, string.Join(",", res.Ns)));
                pyCode.AppendLine(string.Format("experimentos['{0}']['clasico'] = [{1}]", res.Nombre, string.Join(",", res.TClasico.Select(x => x.ToString(ci)))));
                pyCode.AppendLine(string.Format("experimentos['{0}']['strassen'] = [{1}]", res.Nombre, string.Join(",", res.TStrassen.Select(x => x.ToString(ci)))));
                pyCode.AppendLine(string.Format("experimentos['{0}']['is_hybrid'] = {1}", res.Nombre, res.EsExpHibrido ? "True" : "False"));
                
                if (res.EsExpHibrido)
                {
                    pyCode.AppendLine(string.Format("experimentos['{0}']['hib16'] = [{1}]", res.Nombre, string.Join(",", res.THib16.Select(x => x.ToString(ci)))));
                    pyCode.AppendLine(string.Format("experimentos['{0}']['hib32'] = [{1}]", res.Nombre, string.Join(",", res.THib32.Select(x => x.ToString(ci)))));
                    pyCode.AppendLine(string.Format("experimentos['{0}']['hib64'] = [{1}]", res.Nombre, string.Join(",", res.THib64.Select(x => x.ToString(ci)))));
                }
            }

            pyCode.AppendLine(@"
for exp_name, data in experimentos.items():
    ns = data['n']
    plt.figure(figsize=(10, 6))
    
    plt.loglog(ns, data['clasico'], 'o-', linewidth=2, label='Clasico Empirico')
    plt.loglog(ns, data['strassen'], 's-', linewidth=2, label='Strassen Puro')
    
    if data['is_hybrid']:
        plt.loglog(ns, data['hib16'], '^-', linewidth=2, label='Hibrido (n0=16)')
        plt.loglog(ns, data['hib32'], 'v-', linewidth=2, label='Hibrido (n0=32)')
        plt.loglog(ns, data['hib64'], 'd-', linewidth=2, label='Hibrido (n0=64)')
    
    c_clas = data['clasico'][-1] / (ns[-1]**3)
    plt.loglog(ns, [c_clas * (n**3) for n in ns], 'k--', alpha=0.5, label='Teorico O(n^3)')
    
    c_stras = data['strassen'][-1] / (ns[-1]**2.807)
    plt.loglog(ns, [c_stras * (n**2.807) for n in ns], 'r--', alpha=0.5, label='Teorico O(n^2.81)')
    
    plt.xlabel('Tamano de Matriz (n) [Escala Log]')
    plt.ylabel('Tiempo Promedio (segundos) [Escala Log]')
    plt.title('Comparacion C# - ' + exp_name)
    plt.xticks(ns, [str(val) for val in ns])
    plt.legend()
    plt.grid(True, which='both', ls='--', alpha=0.5)
    
    plt.savefig('CSharp_Grafico_' + exp_name + '.png')
    plt.close()
");
            File.WriteAllText("graficador_csharp.py", pyCode.ToString());

            try { Process.Start(new ProcessStartInfo { FileName="python", Arguments="graficador_csharp.py", UseShellExecute=false }).WaitForExit(); }
            catch (Exception) { Console.WriteLine("\nError al llamar a Python. Instale matplotlib."); }
            
            Console.WriteLine("Proceso finalizado con exito.");
        }
    }
}